frontend/catalyst/api_extensions/control_flow.py

```diff
@@ -512,36 +511,78 @@ def _call_with_quantum_ctx(self, ctx):
 
-        out_treedefs, out_signatures = [], []
-        for branch in self.branch_fns + [self.otherwise_fn]:
+        in_sigs, out_sigs = [], []
+        for branch_fn in self.branch_fns + [self.otherwise_fn]:
             quantum_tape = QuantumTape()
+            wfun, in_sig, out_sig = deduce_signatures(
+                branch_fn, [], {}, expansion_strategy=self.expansion_strategy
+            )
+            assert len(in_sig.in_type) == 0
             with EvaluationContext.frame_tracing_context(ctx) as inner_trace:
-                wffa, _, _, out_tree = deduce_avals(branch, [], {})
                 with QueuingManager.stop_recording(), quantum_tape:
-                    res_classical_tracers = [inner_trace.full_raise(t) for t in wffa.call_wrapped()]
-                    res_avals = [shaped_abstractify(res) for res in res_classical_tracers]
-            regions.append(HybridOpRegion(inner_trace, quantum_tape, [], res_classical_tracers))
-            out_treedefs.append(out_tree())
-            out_signatures.append(res_avals)
-
-        _assert_cond_result_structure(out_treedefs)
-        _assert_cond_result_types(out_signatures)
-        out_classical_tracers = [new_inner_tracer(outer_trace, aval) for aval in out_signatures[0]]
-        self._operation = Cond(in_classical_tracers, out_classical_tracers, regions)
-        return tree_unflatten(out_tree(), out_classical_tracers)
-
-    def _call_with_classical_ctx(self):
-        args, args_tree = tree_flatten([])
-        args_avals = tuple(map(_abstractify, args))
-        branch_jaxprs, consts, out_treedefs = initial_style_jaxprs_with_common_consts1(
-            (*self.branch_fns, self.otherwise_fn), args_tree, args_avals, "cond"
+                    res_classical_tracers = [inner_trace.full_raise(t) for t in wfun.call_wrapped()]
+
+            explicit_return_tys = collapse(out_sig.out_type(), res_classical_tracers)
+            hybridRegion = HybridOpRegion(inner_trace, quantum_tape, [], explicit_return_tys)
+            regions.append(hybridRegion)
+            in_sigs.append(in_sig)
+            out_sigs.append(out_sig)
+
+        _assert_cond_result_structure([s.out_tree() for s in out_sigs])
+        _assert_cond_result_types([[t[0] for t in s.out_type()] for s in out_sigs])
+        out_tree = out_sigs[-1].out_tree()
+        all_consts = [s.out_consts() for s in out_sigs]
+        out_types = [s.out_type() for s in out_sigs]
+        # TODO: Creating output tracers like this is not quite cocrect. We want to calculate the
+        # unified result type first (see the call to `unify_convert_result_types`) and only then use
+        # it as the output type of this primitive. But this unification will only happen during the
+        # quantum tracing, when all the user-difined transformations are applied. Since we need to
+        # create output classical tracers here and now, we blindly use out_type of the last cond
+        # branch.
+        out_type_guess = out_types[-1]
```

Comment:
I am not sure what the answer for this will be. Can you create a counter example that you know will fail? Maybe we can just set a test and have it as `pytest.mark.xfail` and fix it in another release? But we should at least know what programs are invalid.

Reply:
Agree. The problem is not that easy to reproduce, [here](https://github.com/PennyLaneAI/catalyst/blob/427d5ad7d623d7913954cbd1df3614d1ca735a71/frontend/test/pytest/test_conditionals.py#L316) is the xfailing test.

Reply:
Mmm... interesting! Thank you! This only fails if we are in a quantum context. What you said in the meeting about tracing quantum components makes more sense now. Thanks!